contrib/hit/lex.h

```diff
@@ -32,14 +33,20 @@ enum class TokType
 /// Token represents an (atomic) token/quantum of input text.
 struct Token
 {
-  Token(TokType t, const std::string & val, size_t offset = 0, int line = 0);
+  Token(TokType t,
+        const std::string & val,
+        const std::shared_ptr<std::string> & name,
+        size_t offset = 0,
+        int line = 0);
   /// str returns a human-friendly string representation of the token.
   std::string str();
 
   /// type identifies the category/type of the token (i.e. String, Number, Comment, etc.)
   TokType type;
   /// val is the actual text from the input that makes this token.
   std::string val;
+  /// name of the original input file
+  std::shared_ptr<std::string> name;
```

Comment:
Why a shared pointer?  Just for performance reasons?  Did you run into cases where it mattered or is this just preventative.  I'd prefer just a vanilla string unless we have confidence that it will be a problem.  And then optimize later if it evolves into a problem.

Reply:
Hmm, fair point. It seemed like a no brainer (due to the large number of tokens and resulting redundant memory usage) and I don't see any obvious drawbacks. If I change everything to vanilla strings that'd be as much work as the (possibly premature) optimization in the other direction would have been, and the code would not be much simpler.

Reply:
I'll still do this and profile it...

Reply:
Parsing a 15kB input 10,000 times:

- Shared pointer: 19.6216s
- Vanilla string: 18.8671s

:-D 

Holy cow... that's ... unexpected!